protocols/mempool/protocol-abstract.md

```diff
@@ -0,0 +1,232 @@
+`peers`, causing the send call to fail or last for a long period would be more
+complex.
+
+The practical effect of scheduling the send of transactions to peers, while
+preserving a per-peer ordering, is to produce multiple _queues_, one per peer,
+of transactions to be sent to that peer.
+The second refining steps comes from the observation that those multiple queues
+would be mostly _identical_, containing the same transactions, with the only
+exception that the queue associated to the peer from which a transaction was
+received for the first time will not include that transaction.
+The multiple queues can then be replaced by a single queue, contain all the
+transactions the node has to relay to peers.
+
+The algorithm below thus include a `mempool` variable, a list of transactions
+that have to be relayed to peers.
+And, for each peer, a `send_routine` is created in order to iterate through the
+`mempool` and send all stored transactions to that peer.
+To handle exceptions of the algorithm, this refined version also includes a
+`senders` map that stores, for every transaction, the set of peers, if any, to
+which that transaction should not be sent.
+This set is initialized to the peer from which a transaction is received via
+the newtork for the first time.
+
+By keeping all the transactions that should be sent to peers in a list, the
+refined algorithm also addresses the addition of new peers to the `peers` set.
+In fact, all transactions in the `mempool` list are sent to that peer, until
+eventually it catches up with the already connected peers.
+From this point, the refined and the original algorithm essentially operates in
+the same way.
+
+Another advantage of keeping all transactions to relay in a single list is the
+possibility of removing from the `mempool` transactions that have been committed.
+While this operation is not really simple to implement in practice, it prevents
+sending committed transactions whose send was originally scheduled, either to
+existing and new peers.
+Note, however, that only adding the transaction to the `seen` map should be
+enough for correctness, in particular because a second check to the `seen` map
+could be performed before actually sending the transaction to a peer.
+
+
+```
+var peers   // other nodes to which we are directly connected
+var seen    // map of transactions already seen
+var mempool // list of transactions
+var senders // map of processes from which transactions were received
+
+upon brodcast(tx):
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        mempool = append(mempool, tx)
+
+upon receive <TX, tx> from p:
+    if not seen[tx] and valid(tx)
+        seen[tx] = true
+        senders[tx] = p
+        mempool = append(mempool, tx)
+    else if seen[tx] // Optional improvement
+        senders[tx] = append(senders[tx], q)
+
+upon commit(txs):
+    for all tx in txs
+        seen[tx] = true
+        mempool = remove(mempool, tx) // Optional improvement
```

Comment:
This is not happening in the reactor.

Reply:
What? We do not remove a committed transaction from the mempool?